### PR TITLE
erasure-code: enable isa-l EC for aarch64 platform

### PR DESCRIPTION
**Pull request by rosinL**
add compile options to support erasure-code on aarch64
platform.

Signed-off-by: Hang Li <lihang48@hisilicon.com>
Signed-off-by: luo rixin <luorixin@huawei.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -123,6 +123,11 @@ if(HAVE_INTEL)
     HAVE_BETTER_YASM_ELF64)
 endif()
 
+if(HAVE_ARM)
+  if(HAVE_ARMV8_SIMD)
```
Comment (tchaikov) [inline src/CMakeLists.txt]: i think we need to update 

https://github.com/ceph/ceph/blob/cb399a9b39f95f352552e155bedfac346ccac0a0/src/CMakeLists.txt#L112-L124

to take `HAVE_ARMV8_SIMD ` into consideration. and then conditionalize on `HAVE_INTEL` and `HAVE_ARMV8_SIMD` in `src/erasure-code/isa/CMakeLists.txt`.

Comment (rosinL) [reply src/CMakeLists.txt]: it is odd for add HAVE_ARMV8_SIMD under HAVE_INTEL. The src/erasure-code/isa/ is added by src/erasure-code/CMakeLists.txt. I think conditionalize on HAVE_INTEL in src/erasure-code/isa/CMakeLists.txt is inappropriate.

Comment (tchaikov) [reply src/CMakeLists.txt]: > it is odd for add HAVE_ARMV8_SIMD under HAVE_INTEL.

it is. and i don't suggest you to do so.

>The src/erasure-code/isa/ is added by src/erasure-code/CMakeLists.txt. I think conditionalize on HAVE_INTEL in src/erasure-code/isa/CMakeLists.txt is inappropriate.

again, i didn't recommend this.

lemme be more specific. you could change 

https://github.com/ceph/ceph/blob/cb399a9b39f95f352552e155bedfac346ccac0a0/src/CMakeLists.txt#L112-L124

to something like

```cmake
if(HAVE_INTEL OR HAVE_ARMV8_SIMD)
  # ...
```

Comment (rosinL) [reply src/CMakeLists.txt]: I am sorry ,I am a bit puzzled about this. I think HAVE_BETTER_YASM_ELF64 is for x86 only, it cann't reuse for aarch64. if we consider HAVE_ARMV8_SIMD here we need HAVE_AARCH64_ASM_ELF64  to identify aarch64, It is like below, and then conditionalize on HAVE_BETTER_YASM_ELF64 and HAVE_AARCH64_ASM_ELF64  in src/erasure-code/isa/CMakeLists.txt
```cmake
if(HAVE_INTEL OR HAVE_ARMV8_SIMD)
  if(HAVE_INTEL)
     # ...
     set HAVE_BETTER_YASM_ELF64 1
   elseif(HAVE_ARMV8_SIMD)
     # ...
     set HAVE_AARCH64_ASM_ELF64 1
    endif()
endif()
```
or we just use HAVE_ARMV8_SIMD instead, there is no change in src/CMakelists.txt and 
modify src/erasure-code/CMakeLists.txt like below  and then conditionalize on HAVE_BETTER_YASM_ELF64 and HAVE_ARMV8_SIMD in src/erasure-code/isa/CMakeLists.txt
```cmake
if (HAVE_BETTER_YASM_ELF64 OR HAVE_ARMV8_SIMD )
  add_subdirectory(isa)
  set(EC_ISA_LIB ec_isa)
endif (HAVE_BETTER_YASM_ELF64 OR HAVE_ARMV8_SIMD )

Comment (tchaikov) [inline src/CMakeLists.txt]: i think `HAVE_ARMV8_SIMD` implies `HAVE_ARM`. so `if(HAVE_ARM)` is not necessary. am i right? and why bother defining `HAVE_AARCH64_ASM_ELF64`? just for the sake of symmetry ?

File: src/erasure-code/CMakeLists.txt [hunk 2]
```diff
@@ -22,10 +22,10 @@ add_subdirectory(lrc)
 add_subdirectory(shec)
 add_subdirectory(clay)
 
-if (HAVE_BETTER_YASM_ELF64)
+if (HAVE_BETTER_YASM_ELF64 OR HAVE_AARCH64_ASM_ELF64)
```
Comment (tchaikov) [inline src/erasure-code/CMakeLists.txt]: s/HAVE_AARCH64_ASM_ELF64/HAVE_ARMV8_SIMD/ .

Comment (tchaikov) [reply src/erasure-code/CMakeLists.txt]: also, please take this opportunity to remove the space after "if".

File: src/erasure-code/CMakeLists.txt [hunk 3]
```diff
@@ -22,10 +22,10 @@ add_subdirectory(lrc)
...
   add_subdirectory(isa)
   set(EC_ISA_LIB ec_isa)
-endif (HAVE_BETTER_YASM_ELF64)
+endif (HAVE_BETTER_YASM_ELF64 OR HAVE_AARCH64_ASM_ELF64)
```
Comment (tchaikov) [inline src/erasure-code/CMakeLists.txt]: could you just put
```c++
endif()
```
here? as the `if` block is very small.

File: src/CMakeLists.txt (hunk 1 above)
Comment (rosinL) [reply src/CMakeLists.txt]: yes, you are right, HAVE_AARCH64_ASM_ELF64 is not nessary, just for corresponding to HAVE_BETTER_YASM_ELF64. we can just use HAVE_ARMV8_SIMD.

**Comment by tchaikov** [comment]
jenkins test make check

File: src/common/options.cc [hunk 4]
```diff
@@ -2597,7 +2597,7 @@ std::vector<Option> get_global_options() {
 
     Option("osd_erasure_code_plugins", Option::TYPE_STR, Option::LEVEL_ADVANCED)
     .set_default("jerasure lrc"
-  #ifdef HAVE_BETTER_YASM_ELF64
+  #if defined(HAVE_BETTER_YASM_ELF64) || defined(HAVE_ARMV8_SIMD)
```
Comment (tchaikov) [inline src/common/options.cc]: you need to add something like
```c++
/* Define it if isa-l is compiled for arm64 */
#cmakedefine HAVE_ARMV8_SIMD
```

**Comment by tchaikov** [comment]
jenkins test make check arm64

File: src/erasure-code/isa/CMakeLists.txt [hunk 5]
```diff
@@ -2,61 +2,90 @@
...
+    ${isal_src_dir}/erasure_code/gf_vect_mad_avx512.asm
+    ErasureCodeIsa.cc
+    ErasureCodeIsaTableCache.cc
+    ErasureCodePluginIsa.cc
+    xor_op.cc
+  )
+elseif(HAVE_ARMV8_SIMD)
+  set(isa_srcs
+    ${isal_src_dir}/erasure_code/ec_base.c
+    ${isal_src_dir}/erasure_code/aarch64/ec_aarch64_highlevel_func.c
```
Comment (tchaikov) [inline src/erasure-code/isa/CMakeLists.txt]:
```
CMake Error at src/erasure-code/isa/CMakeLists.txt:90 (add_library):
  Cannot find source file:

    /home/jenkins-build/build/workspace/ceph-pull-requests-arm64/src/isa-l/erasure_code/aarch64/ec_aarch64_highlevel_func.c
```

Comment (rosinL) [reply src/erasure-code/isa/CMakeLists.txt]: It must to update isa-l(master is best,at least to 02a86df).

Comment (tchaikov) [reply src/erasure-code/isa/CMakeLists.txt]: could you bump the isa-l submodule in a separated commit in this changeset?

Comment (rosinL) [reply src/erasure-code/isa/CMakeLists.txt]: The latest stable release is isa-l 2.28, it doesn't contains ec for aarch64. the next release 2.29 which have ec for aarch64 is coming soon on Jan 2020. Am I update to master or wait for 2.29?

Comment (tchaikov) [reply src/erasure-code/isa/CMakeLists.txt]: i am inclined to hold this change back until isa-l v2.29 rolls out.

Comment (rosinL) [reply src/erasure-code/isa/CMakeLists.txt]: ok, I see.

**Comment by tchaikov** [comment]
unstable.

File: src/erasure-code/isa/CMakeLists.txt (hunk 5 above)
Comment (tchaikov) [reply src/erasure-code/isa/CMakeLists.txt]: https://github.com/intel/isa-l/tree/v2.29.0


@rosinL FYI

Comment (rosinL) [reply src/erasure-code/isa/CMakeLists.txt]: There are some compile errors for ceph on x86 caused by some file changes in isa-l 2.29, to update isa-l to v2.29, is the follow steps right? some advices?
1. a PR update https://github.com/ceph/isa-l to 2.29
2. fix bugs caused  by 2.29
3. update submodule isa-l and commit in this PR
@tchaikov

Comment (tchaikov) [reply src/erasure-code/isa/CMakeLists.txt]: @rosinL i can help on the first step after getting the write access to ceph/isa-l. we can include step 2 and 3 in this PR.

Comment (tchaikov) [reply src/erasure-code/isa/CMakeLists.txt]: @rosinL i just sync'ed ceph/isa-l with upstream. could you continue on with the step 2 and step 3?

Comment (rosinL) [reply src/erasure-code/isa/CMakeLists.txt]: ok，we are working on it

**Comment by rosinL** [comment]
ping @tchaikov

**Comment by tchaikov** [comment]
jenkins test make check arm64

**Comment by tchaikov** [comment]
https://jenkins.ceph.com/job/ceph-pull-requests-arm64/14990

**Comment by tchaikov** [comment]
i reran the "make check" after nuking "build/src/test/crimson/" on the build host , the build and test passed.

**Pull request merged**